### PR TITLE
CALCITE-4434 fix cannot implement 'CASE row WHEN row...'

### PR DESCRIPTION
**Pull request by zinking**
rt

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java [hunk 1]
```diff
@@ -860,6 +861,29 @@ public RexNode convertCall(
...
+
+    //generate relation for `=(Row1,Row2)`
+    if (op.kind == SqlKind.EQUALS) {
+      RexNode expr0 = RexUtil.removeCast(exprs.get(0));
+      RexNode expr1 = RexUtil.removeCast(exprs.get(1));
+      SqlKind expr0Kind = expr0.getKind();
+      SqlKind expr1Kind = expr1.getKind();
+      if (expr0Kind == SqlKind.ROW && expr1Kind == SqlKind.ROW) {
+        RexCall call0 = (RexCall) expr0;
+        RexCall call1 = (RexCall) expr1;
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:873]: Can you give an example for `CASE (deptno, job) WHEN (20, 'CLERK') THEN 1`, what is the `exprs.get(0)` and `exprs.get(1)`?

Is exprs.get(0) `(deptno, job` and  exprs.get(1) `(20, 'CLERK')`

**Review by amaliujia — COMMENTED** [review]
This fix overall looks good to me. Just left a question.

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java [hunk 2]
```diff
@@ -860,6 +861,29 @@ public RexNode convertCall(
             : operandTypeChecker.getConsistency();
     final List<RexNode> exprs = convertOperands(cx, call, consistency);
     RelDataType type = rexBuilder.deriveReturnType(op, exprs);
+
+    //generate relation for `=(Row1,Row2)`
+    if (op.kind == SqlKind.EQUALS) {
+      RexNode expr0 = RexUtil.removeCast(exprs.get(0));
+      RexNode expr1 = RexUtil.removeCast(exprs.get(1));
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:868]: Actually why need to remove cast?

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:868]: If you have to remove cast, then better to add a test that include CAST to show the difference?

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java (hunk 1 above)
Comment (zinking) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:873]: YES

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java (hunk 2 above)
Comment (zinking) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:868]: ROW operation naturally includes a Cast outside. so the example test already has it.

**Comment by julianhyde** [comment]
Although marked as merged, this commit was backed out, then superseded by PR #2309.

**Pull request merged**